operator/pkg/apis/controlplane/v1alpha1/controlplane.go

```diff
@@ -82,3 +84,8 @@ type Instances struct {
 func (c *ControlPlane) ClusterName() string {
 	return c.Name
 }
+
+type ReconcileFinalize interface {
+	Reconcile(context.Context, *ControlPlane) error
+	Finalize(context.Context, *ControlPlane) error
```

Comment:
Looks like finalize doesn't make sense in many/most of your controllers like etcd and addons (kubeproxy and coredns) ? Wondering why you wanted it to be here ?

Reply:
Because I need to pass `ReconcileFinalize` interface for tests else there is no way to provide a fake provider. For some of the controller its a noop.